src/lua/protobuf.lua

```diff
@@ -0,0 +1,723 @@
+--
+-- protocol_def = {
+--     protocol.message(<message_name>, <message_def>),
+--     protocol.enum(<enum_name>, <enum_def>),
+--     <...>
+-- }
+--
+-- Returns a table of the following structure:
+--
+-- protocol = {
+--     ['MessageName_1'] = {
+--         type = 'message'
+--         name = 'MessageName_1'
+--         field_by_name = {
+--             ['FieldName_1'] = <..field_def..>,
+--             ['FieldName_2'] = <..field_def..>,
+--             <...>
+--         },
+--         field_by_id = {
+--             [1] = <..field_def..>,
+--             [2] = <..field_def..>,
+--             <...>
+--         },
+--     },
+--     ['EnumName_1'] = {
+--         type = 'enum'
+--         name = 'EnumName_1'
+--         id_by_value = {
+--             [<string>] = <number>,
+--             [<string>] = <number>,
+--             <...>
+--         },
+--         value_by_id = {
+--             [<number>] = <string>,
+--             [<number>] = <string>,
+--             <...>
+--         },
+--     },
+--     <...>
+--}
+--
+--where <..field_def..> is a table of following structure:
+--
+--field_def = {
+--    type = 'MessageName' or 'EnumName' or 'int64' or <...>,
+--    name = <string>,
+--    id = <number>,
+--    repeated = nil or true,
+--}
+local function protocol(protocol_def)
+    local res = {}
+    -- Declaration table is used to check forward declarations
+    -- false -- this type used as the field type in the message was not defined
+    -- true -- this field type was defined
+    local declarations = {}
+    for _, def in pairs(protocol_def) do
+        if declarations[def.name] then
+            error(('Double definition of name %q'):format(def.name))
+        end
+        if def.type == 'message' then
+            for _, field_def in pairs(def.field_by_id) do
+                local standard = scalars[field_def.type] ~= nil
+                local declared = declarations[field_def.type]
+                if not standard and not declared then
+                    declarations[field_def.type] = false
+                end
+            end
+        end
+        declarations[def.name] = true
+        res[def.name] = def
+    end
+    -- Detects a message or a enum that is used as a field type in message
+    -- but not defined in protocol. Allows a type be defined after usage
+    for def_type, declared in pairs(declarations) do
+        if not declared then
+            error(('Type %q is not declared'):format(def_type))
+        end
+    end
+    return setmetatable(res, protocol_mt)
+end
+
+-- }}} Constructors: message, enum, protocol
+
+
+-- {{{ Global helpers
+
+local function is_number64(value)
+    if type(value) == 'cdata' and (ffi.istype(int64_t, value) or
+        ffi.istype(uint64_t, value)) then
+        return true
+    end
+end
+
+-- Checks 'number' type value to be whole and 'cdata' type value to be
+-- number64 (look function above). Doesn't specifically check type of
+-- input value.
```

Comment:
The 'whole' word means 'entire', not 'integral'. I usually use 'integral' or 'without a fractional part'.

The 'don't specifically check type' phrase is a bit value. I would say 'expects/assumes a value of `number` or `cdata` type as an input'.